src/keri/mod.rs

```diff
@@ -1,71 +1,223 @@
+        Ok(signed)
+    }
+
+    /// Incepts instance of KERI and includes EXTRA keys provided as parameter
+    /// CURRENT Public verification key is extracted directly from KeyManager
+    ///  - it should not be included into `extra_keys` set.
+    /// # Parameters
+    /// * `extra_keys` - iterator over tuples of `(keri::derivation::Basic, keri::keys::Key)`
+    /// # Returns
+    /// `Result<keri::event_message::SignedEventMessage, keri::error::Error>`
+    ///  where `SignedEventMessage` is ICP event including all provided keys + directly fetched
+    ///  verification key, signed with it's private key via KeyManager and serialized.
+    ///
+    pub fn incept_with_extra_keys(&mut self, extra_keys: impl IntoIterator<Item = (Basic, PublicKey)>)
+        -> Result<SignedEventMessage, Error> {
+            let mut keys: Vec<BasicPrefix> = extra_keys
+                .into_iter()
+                .map(|(key_type, key)| key_type.derive(key)).collect();
+            // Signing key must be first
+            keys.insert(0, Basic::Ed25519.derive(self.key_manager.borrow().public_key()));
+            let icp = EventMsgBuilder::new(EventType::Inception)?
+                .with_prefix(self.prefix.clone())
+                .with_keys(keys)
+                .with_next_keys(vec!(Basic::Ed25519.derive(self.key_manager.borrow().next_public_key())))
+                .build()?;
+
+            let signed = icp.sign(PayloadType::MA, vec!(
+                AttachedSignaturePrefix::new(
+                    SelfSigning::Ed25519Sha512,
+                    self.key_manager.borrow().sign(&icp.serialize()?)?,
+                    0
+                )
+            ));
+            let serialized = signed.serialize()?;
+            self.processor.process(signed_message(&serialized).unwrap().1)?;
+            self.prefix = icp.event.prefix;
+
+            Ok(signed)
+    }
+
+    /// Interacts with peer identifier via generation of a `Seal`
+    /// Seal gets added to our KEL db and returned back as `SignedEventMessage`
+    ///
+    pub fn interact(&self, peer: IdentifierPrefix)
+        -> Result<SignedEventMessage, Error> {
+        let next_sn = match self.processor.db.get_kel_finalized_events(&self.prefix) {
+            Some(mut events) => 
+                match events.next_back() {
+                    Some(db_event) => db_event.signed_event_message.event_message.event.sn,
+                    None => return Err(Error::InvalidIdentifierStat)
+                },
+            None => return Err(Error::InvalidIdentifierStat)
+        };
+        let (pref, seal) = match peer {
+            IdentifierPrefix::SelfAddressing(pref) =>
+                (pref.clone(), Seal::Digest(DigestSeal{dig: pref})),
+            _ => return Err(Error::SemanticError("Can interact with SelfAdressing prefixes only".into()))
+        };
+        let event = EventMessage::new(
+            Event::new(
+                self.prefix.clone(),
+                next_sn,
```

Comment:
Shouldn't it be `next_sn + 1`?

Reply:
if it's called `next_sn` it would already be +1 :)

Reply:
IIRC, it's not incremented anywhere in the `interact` function and is the same as `sn` of the last event from kel.

Reply:
oh, I think confusing thing is the name actually. as this is not a key event, but interaction one - should it have different `sn` from original key event we're working against? I thought it should be same for two reasons:

1) It will break continuous key event sequence;

2) It's not that easy to synchronize `sn` generation between different witnesses, and if same `ice` or `rot` have multiple witnesses added - they will all generate same `next_sn`?

Reply:
According to whitepaper (7.24), interaction event is (non-establishment) key event and it can be added to key event sequence.  `sn` for each event in the key event sequence should be unique and monotonically increasing, so I think it should have different `sn` than last event in kel.

I'm not sure about generating interaction events by witnesses, why `sn`s should be synchronized?

Reply:
then each witness will generate `next_sn = sn + 1`, which will be equal for all witnesses receiving same message (multiple interactions with different entities but same sn anyways).
second problem is that you might not always want to share your interaction events with someone different from who you've interacted in that event for privacy reasons - and that will break your kerl parsability and all following events will be out-of-order.

Reply:
Interaction events are first-class citizens of the KEL. They represent commitments to data which the controller wants to include in their KEL, not messages exchanged privately between parties. The `sn` in interaction events should be incremented as in any other. Consensus between witnesses on the latest `sn` is reached by extracting it from the events you have, as each event in a KEL should have a unique `sn` (events with `sn`s which have already appeared are "in the past", they are either recoveries or duplicitous). Private messages should probably not be represented using key events.

Reply:
I've put + 1 for it.
So as I understand from this discussion is that if you have multiple witness grafts in the same message - they should be notified with that message sequentially and with previously updated witness's message included for proper `sn` generation?

1. create graft message
2. send it to witness 1
3. get NTR1 from witness 1
4. send NTR1 to witness 2
5. send graft message to witness 2
6. get NTR2 from witness 2
7. send NTR1 and NTR2 to witness N
8. send graft message to witness N
9. get NTR-N from witness N
... repeat for all

Reply:
I'm still not sure what do you mean by proper `sn` generation. It seems to me that the order in which witnesses are notified with the message and other witnesses receipts shouldn't matter.